File: iOS/SnowTeeth/Effects/Snowflake.swift
//
//  Snowflake.swift
//  SnowTeeth
//
//  Snow particle model
//

import Foundation
import CoreGraphics

enum ParticleState: Equatable {
    case falling
    case stuck(to: String)
    case sliding(direction: CGVector)
    case gone
}

struct Snowflake {
    var position: CGPoint
    var velocity: CGVector
    var size: CGFloat
    var opacity: Float
    var state: ParticleState

    init(position: CGPoint, velocity: CGVector, size: CGFloat, opacity: Float) {
        self.position = position
        self.velocity = velocity
        self.size = size
        self.opacity = opacity
        self.state = .falling
    }

    mutating func update(deltaTime: TimeInterval) {
        switch state {
        case .falling:
            position.x += velocity.dx * deltaTime
            position.y += velocity.dy * deltaTime

        case .stuck:
            // Particles don't move when stuck
            break

        case .sliding(let direction):
            position.x += direction.dx * deltaTime * 100 // slide speed
            position.y += direction.dy * deltaTime * 100

        case .gone:
            break
        }
    }

    func isOffscreen(screenHeight: CGFloat, screenWidth: CGFloat) -> Bool {
        return position.y > screenHeight + size ||
               position.x < -size ||
               position.x > screenWidth + size
    }
}

File: android/app/src/main/java/com/snowteeth/app/effects/Snowflake.kt
package com.snowteeth.app.effects

import android.graphics.PointF

/**
 * Particle states for snow lifecycle
 */
sealed class ParticleState {
    object Falling : ParticleState()
    data class Stuck(val targetId: String) : ParticleState()
    data class Sliding(val directionX: Float, val directionY: Float) : ParticleState()
    object Gone : ParticleState()
}

/**
 * Core snowflake particle model
 * Ported from iOS Swift implementation
 */
data class Snowflake(
    var position: PointF,
    var velocity: PointF,
    val size: Float,
    val opacity: Float,
    var state: ParticleState = ParticleState.Falling
) {
    /**
     * Updates particle position based on current state
     * @param deltaTime Time elapsed since last update in seconds
     */
    fun update(deltaTime: Float) {
        when (val currentState = state) {
            is ParticleState.Falling -> {
                position.x += velocity.x * deltaTime
                position.y += velocity.y * deltaTime
            }
            is ParticleState.Stuck -> {
                // Don't move
            }
            is ParticleState.Sliding -> {
                position.x += currentState.directionX * deltaTime * 100f
                position.y += currentState.directionY * deltaTime * 100f
            }
            is ParticleState.Gone -> {
                // Don't move
            }
        }
    }

    /**
     * Checks if particle has moved offscreen
     */
    fun isOffscreen(screenHeight: Float, screenWidth: Float): Boolean {
        return position.y > screenHeight + size ||
               position.x < -size ||
               position.x > screenWidth + size
    }
}
